refactor!: VC Update (#57)
KILTprotocol/credential-api: readme.md

    /** ID of the message this message responds to */
    inReplyTo?: string

    /** A list of message IDs of previous messages. When this message B is a response to the message A,
     *  B.references = [...A.references, A.inReplyTo] */
    references?: string[]
}
```

#### Quote

```javascript
interface CostBreakdown {
    tax: Record<string, unknown>
    net: number
    gross: number
}

interface Quote {
    cost: CostBreakdown
    currency: string
    timeframe: string
    termsAndConditions: string
}

> ```

Author: as far as I remember, we removed the dedicated signatures. SubmitTerms is signed by the attester and contains the optional Quote:

```typescript
interface SubmitTerms {
    cTypes: ICType[]

    claim: ProposedCredential

    quote?: IQuote
    proof: VCDataIntegrity
}
```

and the RequestAttestation response is also signed:

```typescript
interface RequestAttestation {
    claim: AgreedCredential
    /** quote agreement signed by the claimer
     *  @link https://github.com/KILTprotocol/kilt-extension-api/blob/4c0c2f93958ab72b59b72057a6e9b6aedb5fccef/src/types/Quote.ts#L38 */
    quote?: IQuote
    proof: VCDataIntegrity
}
```

#### Proposed Verifiable Credential

The `ProposedCredential` is a partial `KiltCredentialV1`.

> The attester proposes the content of structure of the credential.

Reviewer: ```suggestion
The structural content of a credential is proposed by an issuer.
```

Since the attester might not yet know the subjects DID, the `credentialSubject.id` property is optional.

> The credential is also not attested, therefore the `proof` and `issuanceDate` properties are missing.

Author: I wouldn't force that all intermediate credentials have to be valid KiltCredentialV1.

```javascript
interface ProposedCredential {
    "@context": [
        "https://www.w3.org/2018/credentials/v1",
        "https://www.kilt.io/contexts/credentials"
    ],
    type: [
        "VerifiableCredential",
        "KiltCredentialV1",

>         "kilt:ctype:0xf0fd09f9ed6233b2627d37eb5d6c528345e8945e0b610e70997ed470728b2ebf"

Author: [7d411c0](https://github.com/KILTprotocol/spec-ext-credential-api/pull/57/commits/7d411c0c99f520ed94f747033c87615c23b55a7f) and [0fea974](https://github.com/KILTprotocol/spec-ext-credential-api/pull/57/commits/0fea9743327551bc15730997970ff35baf077c85)

    ],
    id: string,
    nonTransferable: true,
    credentialSubject: {
        "@context": {

>             "@vocab": "kilt:ctype:0xf0fd09f9ed6233b2627d37eb5d6c528345e8945e0b610e70997ed470728b2ebf#"

Author: [7d411c0](https://github.com/KILTprotocol/spec-ext-credential-api/pull/57/commits/7d411c0c99f520ed94f747033c87615c23b55a7f) and [0fea974](https://github.com/KILTprotocol/spec-ext-credential-api/pull/57/commits/0fea9743327551bc15730997970ff35baf077c85)

        },
        id?: string,
        [key: string]: any
    },
    issuer: string,
    federatedTrustModel: FederatedTrust[],
    credentialStatus: CredentialStatus,
    credentialSchema: {
        id: "ipfs://QmRpbcBsAPLCKUZSNncPiMxtVfM33UBmudaCMQV9K3FD5z",
        type: "JsonSchema2023"
    },
}
```

#### Agreed Verifiable Credential

The `AgreedCredential` is a partial `KiltCredentialV1`.

> It previews the credential that both the claimer and attester agreed upon.

Reviewer: Confirmed based on the PR of Raph. He has moved to issue. I guess this should be informed.

Since the `AgreedCredential` is not attested, it doesn't contain the `issuanceDate` and only a partial `proof`.
The `proof` contains the `type` which is currently limited to `KiltAttestationProofV1` and the `salt` and `commitments`.
Note that the credential subject MUST be present.

> ```javascript

Reviewer 2: js does not have a type system
```suggestion
```typescript
```

Author: > js does not have a type system

not  sure if i can agree with that statement. JS has types so i would argue that it  has a  type  system.  Anyways I changed  it  to typescript

interface AgreedCredential {
    "@context": [
        "https://www.w3.org/2018/credentials/v1",
        "https://www.kilt.io/contexts/credentials"
    ],
    type: [
        "VerifiableCredential",
        "KiltCredentialV1",
        "kilt:ctype:0xf0fd09f9ed6233b2627d37eb5d6c528345e8945e0b610e70997ed470728b2ebf"
    ],

>     id: string,

Author: AFAIK we said that he root hash is already known with the `AgreedCredential`? But the id need to be removed from the `ProposedCredential`. There are no salts and commitments yet.

>     id: string,

Reviewer 2: yeah exactly

>     id: string,

Author: [Removed the status in de3f665](https://github.com/KILTprotocol/spec-ext-credential-api/pull/57/commits/de3f66576771e54e4118b5d7fe8fca53cde56e32) and [the id in 33484c1](https://github.com/KILTprotocol/spec-ext-credential-api/pull/57/commits/33484c12053e9e2e1841d3413b463e31e06b249c)

    nonTransferable: true,
    credentialSubject: {
        "@context": {
            "@vocab": "kilt:ctype:0xf0fd09f9ed6233b2627d37eb5d6c528345e8945e0b610e70997ed470728b2ebf#"
        },
        id: string,
        [key: string]: any
    },
    issuer: string,
    federatedTrustModel: FederatedTrust[],
    credentialStatus: CredentialStatus,
    credentialSchema: {
        id: "ipfs://QmRpbcBsAPLCKUZSNncPiMxtVfM33UBmudaCMQV9K3FD5z",
        type: "JsonSchema2023"
    },
    proof: {
        type: "KiltAttestationProofV1"
        commitments: string[]
        salt: string[]
    }
}
```

#### Credential Status

```javascript
interface CredentialStatus {

>     id: string

Author: I think we will omit the credential Status

>     id: string

Author: [Removed the status in de3f665](https://github.com/KILTprotocol/spec-ext-credential-api/pull/57/commits/de3f66576771e54e4118b5d7fe8fca53cde56e32)

If the attester requires payment to issue this credential, the `quote` MUST be present.
If the attester does not require payment to issue this credential, the `quote` MUST NOT be present.

The communication channel isn't required to provide non-repudiation.
In order for the user to store a proof of the agreed upon terms, the attester MUST sign the `SubmitTerms` object.

> The attached proof MUST follow the [VC Data Integrity specification](https://www.w3.org/TR/vc-data-integrity/#proofs).

Author: I would propose that we add a `SHOULD support the crypto suite "name of the crypto suite we will use"`. Which crypto suite do we use?

> The attached proof MUST follow the [VC Data Integrity specification](https://www.w3.org/TR/vc-data-integrity/#proofs).

Reviewer 2: We got three in the sdk, one for Sr25519 (sr25519-jcs-2023), one for ed25519 (eddsa-jcs-2022) and one for secp256k1 keys (ES256K-jcs-2023 or so)

> The attached proof MUST follow the [VC Data Integrity specification](https://www.w3.org/TR/vc-data-integrity/#proofs).

Author: What to you think about this: [33484c1](https://github.com/KILTprotocol/spec-ext-credential-api/pull/57/commits/33484c12053e9e2e1841d3413b463e31e06b249c)

> The attached proof MUST follow the [VC Data Integrity specification](https://www.w3.org/TR/vc-data-integrity/#proofs).

Reviewer 2: just double-checked, that's indeed what they are called 
<img width="231" alt="image" src="https://github.com/KILTprotocol/spec-ext-credential-api/assets/39338561/8529d1da-dadc-4869-9322-4cd883f2cf9b">

     *  @link https://kiltprotocol.github.io/sdk-js/interfaces/types_src.ICType.html */

>     cTypes: ICType[]

Reviewer: This should be the following:

`
```suggestion
    cTypes: Array<{
        cType: PartialClaim
    }>;
```

Reviewer: However, this should be adapted to the changes made by @rflechtner

Author: Could you elaborate a bit? Why would you change it like that, which changes done by Raphael where?

    claim: ProposedCredential

    /** optional attester-signed binding
     *  @link https://github.com/KILTprotocol/kilt-extension-api/blob/4c0c2f93958ab72b59b72057a6e9b6aedb5fccef/src/types/Quote.ts#L30 */
    quote?: IQuote
    proof: VCDataIntegrity
}
```

> #### 2.a Extension requests credential

Reviewer: I guess it was not to 2.b.

The extension MUST only send the request with active consent of the user.
This is the first step where the user’s DID is revealed to the dApp.

> The previous message in the flow - `'submit-terms'` - contains the `claim` with an optional `subject` field containing a DID URI.

Author: yeah I think I'm reffering to the `SubmitTerms` and the `/claim/credentialSubject/id` property?

> The previous message in the flow - `'submit-terms'` - contains the `claim` with an optional `subject` field containing a DID URI.

Reviewer 2: I think so too, would make sense

> The previous message in the flow - `'submit-terms'` - contains the `claim` with an optional `subject` field containing a DID URI.

Author: [hope this makes more sense](b3c2d61e668c56de6e3f22ac0387e58c2f810ce2)

This `subject` value being provided means that the attester is willing to issue the credential for this specific DID.

> If the `'submit-terms'` message included an unknown DID or none at all as `subject`, the extension MUST ask the user to choose the DID for which the credential will be issued.

Author: That is taken from the old spec.

> If the `'submit-terms'` message included an unknown DID or none at all as `subject`, the extension MUST ask the user to choose the DID for which the credential will be issued.

Reviewer 2: I don't think we need to support shitty old apps in a new extension api version. I don't see why they wouldn't be able to just not pass any DID if they don't want to make a choice.

> If the `'submit-terms'` message included an unknown DID or none at all as `subject`, the extension MUST ask the user to choose the DID for which the credential will be issued.

Author: Its breaking anyways and requires some updates so I agree

> If the `'submit-terms'` message included an unknown DID or none at all as `subject`, the extension MUST ask the user to choose the DID for which the credential will be issued.

Reviewer 2: exactly

> If the `'submit-terms'` message included an unknown DID or none at all as `subject`, the extension MUST ask the user to choose the DID for which the credential will be issued.

Author: [hope this makes more sense](b3c2d61e668c56de6e3f22ac0387e58c2f810ce2)

> If the `'submit-terms'` message included an unknown DID or none at all as `subject`, the extension MUST ask the user to choose the DID for which the credential will be issued.

Reviewer 2: yes that's great!

Otherwise, the extension SHOULD NOT offer the choice, but still MUST get the user’s consent to use this DID.

The extension MUST generate the `salt` values according to the [KiltAttestationProofV1 specification](https://github.com/KILTprotocol/spec-KiltCredentialV1/blob/main/ProofTypes/KiltAttestationProofV1.md#salt) and provide them in the `request-attestation` message.
The extension MUST provide the `delegationId` if the attester provided one in the previous step.

The extension used a temporary DID for the communication channel.
Since this DID is only used for a single interaction, it is not possible for the attester to verify the identity of the claimer until this step.
In order for the attester to hold a proof that the claimer agreed to the terms, the claimer MUST sign the `RequestAttestation` object using the DID specified in the credential subject (field `claim.credentialSubject.id`).
The attester SHOULD store the signed `RequestAttestation` object in case it is needed for later dispute resolution.

> The chosen or confirmed DID URI will be submitted as the `subject` field of the `claim` in the `'request-attestation'` message.

Reviewer 2: again, I don't think there is a `subject`

Author: [fixed in this commit](315037d22e3539d6620f4b6d2c616fd200d0d068) hope the JSON pointer is ok? Not sure it should be relative to the message object or the content of the message. So `/claim/credentialSubject/id` or `/body/content/claim/credentialSubject/id`

Reviewer 2: I think people will understand

The attester MUST only issue a credential to this DID.
The attester MUST use the provided `salt` values for the proof of the credential.
The attester MAY reject the request if this DID is different from the `subject` in the previous `'submit-terms'` message.

|              |                         |
| ------------ | ----------------------- |
| direction    | `extension -> dApp`     |
| message_type | `'request-attestation'` |

```typescript
interface RequestAttestation {
    claim: AgreedCredential

>     /** quote agreement signed by the claimer

Reviewer 2: not signed, right?

Author: [not signed and also not the IQuote from the SDK but the Quote defined in this doc](8a291effde4d2dc59e32c2d4189a6c2b4fbf9db2)

}
```

#### 4.b Optional: Extension rejects payment

The extension MUST send a [rejection message](#rejections) if the user cancels or rejects the payment.

> #### 5.a Attester submits credential

Reviewer 2: I think we should also allow sending only this message to allow scenarios where everything from entering and confirming claim data to selecting a DID to be used (ideally also requesting a challenge to be signed with that DID) is handled in an app outside the the credential api flow, and only the resulting credential is submitted to the extension.

Author: I'm not 100% sure that this is enough: [8787eae](https://github.com/KILTprotocol/spec-ext-credential-api/pull/57/commits/8787eae5fc46fc886a67533709fece0df5baac54).
Maybe we need to specify more? I thought about how the salts are generated. But maybe we leave all that open since this would be the most flexible for the issuer? Maybe at this point we also need to specify what are valid combinations of messages? E.g. it's not valid to request payment without submitting terms before that.

Reviewer 2: I think it makes sense to add detail on who generates what parts of the credential. If there is an agreement, the issuer must base the credential they issue on the agreed upon data, including salts (if given?); If there is no agreement and the submit-credential is the only step, then the issuer creates all that data on their side, including salts.

Repeat for multiple required credentials.

#### 1. DApp or extension requests credential

|              |                        |
| ------------ | ---------------------- |
| direction    | `dApp <-> extension`   |
| message_type | `'request-credential'` |

> Multiple CTypes MAY be requested here only if they can be used interchangeably.

Author: Do we want to make it impossible to request a credential with any CType? If yes why?

> Multiple CTypes MAY be requested here only if they can be used interchangeably.

Reviewer 2: the only use case for requesting any credential is a verifier demo app; any actual use case would require a specific type of credential

> Multiple CTypes MAY be requested here only if they can be used interchangeably.

Author: But is there a reason to require a ctype? I don't think it's a good idea to make this a MUST only because we can't think of a use case without this.

> Multiple CTypes MAY be requested here only if they can be used interchangeably.

Reviewer 2: Fair, especially bc it can be useful for demo apps

            /** optional list of DIDs of attesters trusted by this verifier */
            trustedAttesters?: string[]

            /**
             * list of credential attributes which MUST be included when submitting the credential.

>              * The properties are identified using JSONPath.

Reviewer 2: let's do json pointers, which are simpler and more specific

Author: fixed in https://github.com/KILTprotocol/spec-ext-credential-api/commit/7d37be367d3ef90811deb76b644782f598834d41

             */
            requiredProperties: string[]
        }
    ]

    /** Optional DID URI the credential should be issued to */
    subject?: string

    /** 24 random bytes as hexadecimal */
    challenge: string
}
```

##### Required Properties

The `cTypes.requiredProperties` property specifies which fields of the credential MUST be included in the presentation.
All fields which are not covered by the selective disclosure scheme are mandatory and MUST not be listed in the `requiredProperties`.

#### 2. Extension or dApp sends credential

The extension MUST only send the credential with active consent of the user.
This is the first step where the user’s DID is revealed to the dApp.

> The `challenge` from the previous message MUST be used to create the [verifiable presentation](https://www.w3.org/TR/vc-data-model/#presentations-0)

Reviewer 2: we could add that this would be added to the data integrity proof on its `challenge` field
